### DISCUSSION TITLE
Optimize EstimateDistances computations

### DISCUSSION THREAD

**Discussion by aghozlane** [cogent3/cogent3#2134] [Q&A] [answered]
Dear cogent3 developpers,

We are performing a lot of benchmark around cogent3 and I am looking for ways to optimize my code.
Here we measure the F81, GTR and K80 distances on aligned sequences and compute the upgma tree:
<img width="999" alt="image" src="https://github.com/user-attachments/assets/d44dcf76-d07e-4e77-8249-321d37c9ff0e">

```
aligned_seqs = load_aligned_seqs(input_fasta, moltype = "dna")
d = EstimateDistances(aligned_seqs, submodel = GTR()) # or F81() or K80()
d.run(show_progress=False)
mycluster = upgma(d.get_pairwise_distances())
mycluster = mycluster.unrooted_deepcopy()
```

Is there anyway to speed up these few lines of code ?

Besides being slower, we are really happy of the result that we obtain.

Best regards,
Amine

**Comment by GavinHuttley** [comment]
Hi @aghozlane !
The short answer is yes, I think the speed can be improved a lot.

We have just merged in a major refactor of the core objects for sequences and alignments (we will be making a release at the end of this week). We have found that these changes provide a major speed boost for a selected set of operations we have examined. The new approach to alignment handling also paves the way for vastly more efficient parallel computation.

At present, distance calculations do not yet take advantage of them. I would predict we can get close to an order of magnitude speed improvement for the subset of models with exact (i.e. analytical) solutions.

If you are interested, I can sketch out how to refactor our algorithms to deliver that speed and you could have a go at doing it.

On another note, the TN93 model is quite close in form to GTR but has analytical solutions and is thus must faster. Also, if the data exhibits heterogeneity in nucleotide composition, one should use the paralinear distance.

**Reply by GavinHuttley** [reply]
Quick comment, I just benchmarked a revised JC69 calculator, it can do all pairwise distances between 1k by 30k length sequences in < 2 seconds.

**Reply by aghozlane** [reply]
Thank you for all the updates. I confirm that in our benchmark TN93 is really fast and is good approximation of the GTR.

How can I access to your revised JC69 calculator ?

**Reply by GavinHuttley** [reply]
Hi @aghozlane ,
We have just released the major revision I mentioned above. I have included the implementation I suggested into that release. While it is tested, there could be bugs! In which case, please report.

To get the revised version `pip install cogent3 -U`. Then you can do the following
```python
aligned_seqs = load_aligned_seqs(input_fasta, moltype = "dna", new_type=True) # note new_type arg
dists = aligned_seqs.distance_matrix(calc="jc69", parallel=True)
```
> **Note**
> These new distance calculators are not hooked up to the older Alignment classes, so you have to set `new_type=True`.

The number of distance measures via this interface is reduced to jc69, tn93 and paralinear. (Parallel is off by default.)

On another note, the returned distance matrix object can build a NJ tree for you as
```python
tree = dists.quick_tree()
```
It would be great to see the revised performance plot.

**Reply by aghozlane** [reply]
And here is it:
<img width="1021" alt="image" src="https://github.com/user-attachments/assets/42ed2d1a-2499-4549-9ca5-f08c2083b749">
and here the zoom on the parallel
<img width="610" alt="image" src="https://github.com/user-attachments/assets/f0452c33-4511-4f95-91bc-b00d8b2b7430">

The TN93 works really fine to distinguish our sequences and is really fast to compute with your approach.
We will have much more sequence with the real dataset, so I guess the parallel version will have more impact with larger dataset. We can see here the curve is flat because our benchmark does not contain enough sequence.
Let me know if we can do better :)

**Reply by GavinHuttley** [reply, marked as answer]
Hi @aghozlane ,
Thanks for the updated plots. Awesome to see all the cogent3 teams hard work is leading to some massive performance gains!

Single CPU performance may get some additional gains by trying to take advantage of SIMD (I am unsure how difficult that will be). But I think you need to do profiling on the type of data that is closer to what you will actually be getting.

There's a notebook at [this repo](https://github.com/cogent3/cogent3_and_plotly) (which was used for [this blog post](https://plotly.com/blog/decoding-genomes-cogent3-plotly/)). You can edit that code to get a download of ~90k whole genomes of SARS-COV-2 (the alignment is ~30kb long).

On my Apple M2 laptop, I can get all pairwise distances for 10k of those genomes in 27seconds running in parallel, 2mins 26 seconds in serial.

But if you will have many such alignments, it may be best to parallelise at the level of alignment, rather than calculations for a single alignment.

So your optimal solution will depend on your data.
